### ISSUE TITLE
Charset conversion mysql-connector-j-8.0.31.jar

### ISSUE THREAD

**Issue by djamadeus** [exasol/mysql-virtual-schema#26]
When reading data from virtual schema I get a converison-error from UTF8 to ASCII,. Reading directly from the jdbc-connection works though.

EXA_DB.VIRTUAL_JOINTDB_OFFICE> Select * from(
                               IMPORT  FROM JDBC
                               AT MYSQL_JOINTDB_JDBC_CONNECTION
                               STATEMENT 'SELECT * from office.erstattungen_positionen limit 100')
[2022-11-16 10:13:59] 100 rows retrieved starting from 1 in 1 s 85 ms (execution: 1 s 53 ms, fetching: 32 ms)
EXA_DB.VIRTUAL_JOINTDB_OFFICE> Select * from virtual_jointDB_office."erstattungen_positionen" limit 100
[2022-11-16 10:21:58] [42636] ETL-3009: [Column=23 Row=0] [Charset conversion from 'UTF-8' to 'ASCII' failed [Invalid or incomplete multibyte or wide character]] (Session: 1749641517751205888).

On the Source-DB the charset is latin1 (yes i know :(, not my fault^^).

**Comment by djamadeus** [comment]
the column contains german letters like "Ü". The Problem here seems to be, that the virtual schema adapter gets latin1 fromt he column definition so it expects latin1, the mysql JDBC driver already converts this to utf-8 though.

**Comment by djamadeus** [comment]
Jep switching back to 9.0.5 of the adapter script solved the issue for me as well

**Comment by ckunki** [comment]
To summarize: version `virtual-schema-dist-9.0.5-mysql-3.0.0.jar` did not show the behavior described above while `virtual-schema-dist-10.0.1-mysql-4.0.0.jar` does so, right?

Hi @djamadeus could you please add the result of describing the table? 

In Exasol DB SQL you get this from `DESCRIBE <schema>.<table-name>;` with `<schema>` being the name of the virtual schema you created and `<table-name>` replaced by the name of your table, in total probably `office.erstattungen_positionen`, then.

After that I would like to exclude potential effects due to caching the schema description. We can do so by refreshing the virtual schema with `ALTER VIRTUAL SCHEMA <schema> REFRESH;` followed by an additional `DESCRIBE <schema>.<table-name>;`, see also https://docs.exasol.com/db/latest/sql/alter_schema.htm.

Based on that I will add more information.

**Comment by djamadeus** [comment]
@ckunki yes that only happens with version 10.0.1
Descripttion-result with 9.0.5-adapter:

| COLUMN\_NAME | SQL\_TYPE |
| :--- | :--- |
| id | DECIMAL\(10,0\) |
| datum | TIMESTAMP |
| datum\_ursprung | TIMESTAMP |
| userid | VARCHAR\(16\) ASCII |
| orderid | VARCHAR\(32\) ASCII |
| rekla\_id | VARCHAR\(10\) ASCII |
| quelle | CHAR\(15\) ASCII |
| quelle\_id | VARCHAR\(15\) ASCII |
| art | CHAR\(19\) ASCII |
| grund | VARCHAR\(25\) ASCII |
| mitarbeiter | VARCHAR\(50\) ASCII |
| bemerkung | VARCHAR\(160\) ASCII |
| artnr | VARCHAR\(25\) ASCII |
| sellid | VARCHAR\(15\) ASCII |
| bezeichnung | VARCHAR\(50\) ASCII |
| bezeichnung\_zusatz | VARCHAR\(50\) ASCII |
| anzahl | DECIMAL\(10,0\) |
| einzelpreis | DECIMAL\(8,2\) |
| brutto\_netto | CHAR\(6\) ASCII |
| steuersatz | DECIMAL\(4,1\) |
| datum\_bestellung | TIMESTAMP |
| zahlungsart\_bestellung | VARCHAR\(32\) ASCII |
| prioritaet | CHAR\(6\) ASCII |
| status | CHAR\(11\) ASCII |
| status\_bemerkung | VARCHAR\(160\) ASCII |
| id\_gutschrift | DECIMAL\(10,0\) |

**Comment by djamadeus** [comment]
do you want me to test the refresh-flag with the 10 or 9.x adapter?

**Comment by ckunki** [comment]
Thank you very much for the additional information. Please apply the refresh for the new version 10.
Could please also tell me the version of the Exasol database you are using?

**Comment by ckunki** [comment]
Command `DESCRIBE <schema>.<table-name>;` shows that Exasol DB expects ASCII, which matches the error message above: 
`ETL-3009: [Charset conversion from 'UTF-8' to 'ASCII' failed [Invalid or incomplete multibyte or wide character]]`. The prefix `ETL-` of the error code indicates the message being generated by the ExaLoader which is a core component of the database. 

Could you help me to get insight into the perspective of the virtual schema adapter (another component different from ExaLoader)? 

You can do so by using `EXPLAIN VIRTUAL <sql command>;`, see https://docs.exasol.com/db/latest/sql/explain_virtual.htm.
Please use the same SQL command that created the initial error message.

`EXPLAIN VIRTUAL` returns a table and if you could attach column "PUSHDOWN_SQL" to the current ticket, that will hopefully show which component indicated a UTF-8 encoding, which probably is incorrect and therefore cannot be converted to ASCII.

**Comment by djamadeus** [comment]
`
CREATE VIRTUAL SCHEMA virtual_jointDB_office_test
    USING system_adapter_scripts.ADAPTER_SCRIPT_MYSQL
    WITH
    CONNECTION_NAME = 'MYSQL_JOINTDB_JDBC_CONNECTION'
    CATALOG_NAME = 'office';


Select * from virtual_jointDB_office_test."erstattungen_positionen";
[42636] ETL-3009: [Column=23 Row=0] [Charset conversion from 'UTF-8' to 'ASCII' failed [Invalid or incomplete multibyte or wide character]] (Session: 1750109444829478912)


ALTER VIRTUAL SCHEMA virtual_jointDB_office_test REFRESH;

Select * from virtual_jointDB_office_test."erstattungen_positionen";
[Column=23 Row=0] [Charset conversion from 'UTF-8' to 'ASCII' failed [Invalid or incomplete multibyte or wide character]] (Session: 1750109444829478912)

explain virtual Select * from virtual_jointDB_office_test."erstattungen_positionen";
`
Pushdown_SQL:
`IMPORT INTO (c1 DECIMAL(10, 0), c2 TIMESTAMP, c3 TIMESTAMP, c4 VARCHAR(16) ASCII, c5 VARCHAR(32) ASCII, c6 VARCHAR(10) ASCII, c7 CHAR(15) ASCII, c8 VARCHAR(15) ASCII, c9 CHAR(19) ASCII, c10 VARCHAR(25) ASCII, c11 VARCHAR(50) ASCII, c12 VARCHAR(160) ASCII, c13 VARCHAR(25) ASCII, c14 VARCHAR(15) ASCII, c15 VARCHAR(50) ASCII, c16 VARCHAR(50) ASCII, c17 DECIMAL(10, 0), c18 DECIMAL(8, 2), c19 CHAR(6) ASCII, c20 DECIMAL(4, 1), c21 TIMESTAMP, c22 VARCHAR(32) ASCII, c23 CHAR(6) ASCII, c24 CHAR(11) ASCII, c25 VARCHAR(160) ASCII, c26 DECIMAL(10, 0)) FROM JDBC AT MYSQL_JOINTDB_JDBC_CONNECTION STATEMENT 'SELECT `erstattungen_positionen`.`id`, `erstattungen_positionen`.`datum`, `erstattungen_positionen`.`datum_ursprung`, `erstattungen_positionen`.`userid`, `erstattungen_positionen`.`orderid`, `erstattungen_positionen`.`rekla_id`, `erstattungen_positionen`.`quelle`, `erstattungen_positionen`.`quelle_id`, `erstattungen_positionen`.`art`, `erstattungen_positionen`.`grund`, `erstattungen_positionen`.`mitarbeiter`, `erstattungen_positionen`.`bemerkung`, `erstattungen_positionen`.`artnr`, `erstattungen_positionen`.`sellid`, `erstattungen_positionen`.`bezeichnung`, `erstattungen_positionen`.`bezeichnung_zusatz`, `erstattungen_positionen`.`anzahl`, `erstattungen_positionen`.`einzelpreis`, `erstattungen_positionen`.`brutto_netto`, `erstattungen_positionen`.`steuersatz`, `erstattungen_positionen`.`datum_bestellung`, `erstattungen_positionen`.`zahlungsart_bestellung`, `erstattungen_positionen`.`prioritaet`, `erstattungen_positionen`.`status`, `erstattungen_positionen`.`status_bemerkung`, `erstattungen_positionen`.`id_gutschrift` FROM `office`.`erstattungen_positionen`'`

**Comment by djamadeus** [comment]
when I run the pushdown command:
`IMPORT INTO (c1 DECIMAL(10, 0), c2 TIMESTAMP, c3 TIMESTAMP, c4 VARCHAR(16) ASCII, c5 VARCHAR(32) ASCII, c6 VARCHAR(10) ASCII, c7 CHAR(15) ASCII, c8 VARCHAR(15) ASCII, c9 CHAR(19) ASCII, c10 VARCHAR(25) ASCII, c11 VARCHAR(50) ASCII, c12 VARCHAR(160) ASCII, c13 VARCHAR(25) ASCII, c14 VARCHAR(15) ASCII, c15 VARCHAR(50) ASCII, c16 VARCHAR(50) ASCII, c17 DECIMAL(10, 0), c18 DECIMAL(8, 2), c19 CHAR(6) ASCII, c20 DECIMAL(4, 1), c21 TIMESTAMP, c22 VARCHAR(32) ASCII, c23 CHAR(6) ASCII, c24 CHAR(11) ASCII, c25 VARCHAR(160) ASCII, c26 DECIMAL(10, 0)) FROM JDBC AT MYSQL_JOINTDB_JDBC_CONNECTION STATEMENT 'SELECT erstattungen_positionen.id, erstattungen_positionen.datum, erstattungen_positionen.datum_ursprung, erstattungen_positionen.userid, erstattungen_positionen.orderid, erstattungen_positionen.rekla_id, erstattungen_positionen.quelle, erstattungen_positionen.quelle_id, erstattungen_positionen.art, erstattungen_positionen.grund, erstattungen_positionen.mitarbeiter, erstattungen_positionen.bemerkung, erstattungen_positionen.artnr, erstattungen_positionen.sellid, erstattungen_positionen.bezeichnung, erstattungen_positionen.bezeichnung_zusatz, erstattungen_positionen.anzahl, erstattungen_positionen.einzelpreis, erstattungen_positionen.brutto_netto, erstattungen_positionen.steuersatz, erstattungen_positionen.datum_bestellung, erstattungen_positionen.zahlungsart_bestellung, erstattungen_positionen.prioritaet, erstattungen_positionen.status, erstattungen_positionen.status_bemerkung, erstattungen_positionen.id_gutschrift FROM office.erstattungen_positionen limit 100'
[2022-11-23 12:30:26] [42636] ETL-3009: [Column=23 Row=0] [Charset conversion from 'UTF-8' to 'ASCII' failed [Invalid or incomplete multibyte or wide character]] (Session: 1750109444829478912)
`

**Comment by djamadeus** [comment]
If I counted right, the proble is ` erstattungen_positionen.prioritaet` wich is an enum.
Probably always using UTF8 for enums in the newer adapter?

**Comment by djamadeus** [comment]
descriptiontable with vs-10.0.1-adapter:

| COLUMN\_NAME | SQL\_TYPE | NULLABLE | DISTRIBUTION\_KEY | PARTITION\_KEY |
| :--- | :--- | :--- | :--- | :--- |
| id | DECIMAL\(10,0\) | NULL | NULL | NULL |
| datum | TIMESTAMP | NULL | NULL | NULL |
| datum\_ursprung | TIMESTAMP | NULL | NULL | NULL |
| userid | VARCHAR\(16\) ASCII | NULL | NULL | NULL |
| orderid | VARCHAR\(32\) ASCII | NULL | NULL | NULL |
| rekla\_id | VARCHAR\(10\) ASCII | NULL | NULL | NULL |
| quelle | CHAR\(15\) ASCII | NULL | NULL | NULL |
| quelle\_id | VARCHAR\(15\) ASCII | NULL | NULL | NULL |
| art | CHAR\(19\) ASCII | NULL | NULL | NULL |
| grund | VARCHAR\(25\) ASCII | NULL | NULL | NULL |
| mitarbeiter | VARCHAR\(50\) ASCII | NULL | NULL | NULL |
| bemerkung | VARCHAR\(160\) ASCII | NULL | NULL | NULL |
| artnr | VARCHAR\(25\) ASCII | NULL | NULL | NULL |
| sellid | VARCHAR\(15\) ASCII | NULL | NULL | NULL |
| bezeichnung | VARCHAR\(50\) ASCII | NULL | NULL | NULL |
| bezeichnung\_zusatz | VARCHAR\(50\) ASCII | NULL | NULL | NULL |
| anzahl | DECIMAL\(10,0\) | NULL | NULL | NULL |
| einzelpreis | DECIMAL\(8,2\) | NULL | NULL | NULL |
| brutto\_netto | CHAR\(6\) ASCII | NULL | NULL | NULL |
| steuersatz | DECIMAL\(4,1\) | NULL | NULL | NULL |
| datum\_bestellung | TIMESTAMP | NULL | NULL | NULL |
| zahlungsart\_bestellung | VARCHAR\(32\) ASCII | NULL | NULL | NULL |
| prioritaet | CHAR\(6\) ASCII | NULL | NULL | NULL |
| status | CHAR\(11\) ASCII | NULL | NULL | NULL |
| status\_bemerkung | VARCHAR\(160\) ASCII | NULL | NULL | NULL |
| id\_gutschrift | DECIMAL\(10,0\) | NULL | NULL | NULL |

**Comment by djamadeus** [comment]
Exasol  version is 7.1.15 btw

**Comment by ckunki** [comment]
Thank you very much for all the information. I am confident that this will enable us to find the correct answer and will come back in short.

**Comment by djamadeus** [comment]
jep.
Using this as adapter on same DB works without problems:
`CREATE OR REPLACE JAVA ADAPTER SCRIPT system_adapter_scripts.ADAPTER_SCRIPT_MYSQL_ASCII AS
    %scriptclass com.exasol.adapter.RequestDispatcher;
    %jar /buckets/bucketfs1/bucket1/virtual-schema-dist-9.0.5-mysql-3.0.0.jar;
    %jar /buckets/bucketfs1/bucket1/mysql-connector-j-8.0.31.jar;
/
;`
while this creates the reported behavior:
`CREATE OR REPLACE JAVA ADAPTER SCRIPT system_adapter_scripts.ADAPTER_SCRIPT_MYSQL AS
    %scriptclass com.exasol.adapter.RequestDispatcher;
    %jar /buckets/bucketfs1/bucket1/virtual-schema-dist-10.0.1-mysql-4.0.0.jar;
    %jar /buckets/bucketfs1/bucket1/mysql-connector-j-8.0.31.jar;
/
;
`
IN the changes to version 10 it says something about, not relying on analyzing the received data anymore but using the table meta-information instead...
That sounds a lot like the actual culprit to me ;)

**Comment by ckunki** [comment]
> changes to version 10 ... using the table meta-information instead...
> That sounds a lot like the actual culprit to me ;)

Thanks for mentioning that as this constantly goes through my mind, too.
I only now found out that this already has been backported to Exasol version 7.1.14. So yes, this is still a candidate.

What I am still wondering about: How could the information be different from what we get from `DESCRIBE TABLE`?
And actually we must assume that the meta-information includes an indication for UTF-8 while all the other places don't. Not mentioning any encoding would mean the same as UTF-8 is the default.

Would it be OK to ask for two additional log samples?
1. from the data base / ExaLoader
2. from the virtual schema and the meta-information

For (1) the relevant log files have names like `*EtlJdbc_<etl_proc_num>_<job_num>*`.
Here is an example for mapping a JDBC data type to Exasol:
```
2022-11-22 23:07:42.438 debu: Mapped Column: types[0]=93, internalTypes[0]=INTERNALTYPES_TIMESTAMP
```

For (2) see https://docs.exasol.com/db/latest/database_concepts/virtual_schema/logging.htm

**Comment by ckunki** [comment]
Sorry for assuming you might be able to access these logs. In the meantime I could do some deeper experiments and I am close to a proof of our current assumption which is matching your initial diagnosis.

VSMYSQL version 3.x inspected only the data to detect the data type and got UTF-8 which is accepted by Exasol DB when importing UTF-8 encoded values.

VSMYSQL version 4.x inspects the metadata of the JDBC connection to detect the data type of each database column. In our case this is probably CHAR with octet length 1, i.e. 1 byte per character which is more or less correct for encoding latin1. VSMYSQL maps this to Exasol data type ASCII.

When retrieving actual values in a select statement the JDBC driver converts these character to UTF-8 and Exasol database reports an error as it cannot convert these data to data type ASCII.

![image](https://user-images.githubusercontent.com/81821845/203831007-b99965af-958c-435b-b756-5d3c3f17c654.png)

The question is now how to overcome this. 
I see the following options 
* continue to use the old version of the VSMYSQL 
* say that the German umlaut "Ü" is not strict "ASCII" and recommend to use UTF-8 encoding in MySQL
* enhance VSMYSQL to support a parameter to always assume UTF-8 
* enhance VSMYSQL to support a parameter to ignore connection metadata for data type detection 
* Find a way to suppress JDBC driver to (silently) convert values to UTF-8

I will ponder these options and discuss with my colleagues. 
Please tell me if you have a preference or if you see additional options.

**Comment by djamadeus** [comment]
Since detecting the type has worked flawlessly in the past, maybe having a
parameter to fall back to the 3.x behaviour is best.
I don't know the background why this was changed. Were there other issues
with the old way, or is it "just" performance related?

Thx for your efforts btw!

Stefan Ladenthin CTO
Cell: +49 152 2235 9819
Wunderdata GmbH
Angermünder Str. 20 Haus B 104
16247 Joachimsthal

**Comment by ckunki** [comment]
Thanks for contact information and feeback.

There were a number of reasons for the change in VSMYSQL:
* Improved performance of queries to virtual schema by avoiding one query for each push down
* Enhanced accuracy of data type mapping
* Makes data type mapping simpler and easier to extend
* Fixes wrong data types in some use cases
* Supports additional use cases and correct data types in special situations

Also guessing the data types from the values in the result set appeared to be inappropriate and risky. Although, we must must admit, that in your special case the old behavior seems to be preferable. 😃 

In the meanwhile we did some experiments with settings of the connection which did not change the behavior.

The bug filed for MySQL Connector J: https://bugs.mysql.com/bug.php?id=69328, submitted: 27 May 2013, status Verified i.e. still open. ☹️ further narrows down the options.

In summary we are currently focusing option (O4) as proposed by you.

**Comment by ckunki** [comment]
To ensure correct understanding of the problem I investigated the behavior of the MySQL JDBC driver:

```sql
CREATE TABLE T (c1 CHAR(1), c2  ENUM('A', 'Ü')) CHARACTER SET latin1;
INSERT INTO T VALUES ('Ü', 'Ü');
```

The following results have been retrieved by using the MySQL JDBC driver `com.mysql:mysql-connector-j:8.0.31`.

Column Metadata for MySQL table 'T' provided by JDBC driver:
- Column ‘c1': from adapter notes:` { JdbcDataType = 1 (CHAR), typeName = 'CHAR' }`, type = `CHAR(1) ASCII`.
- Column 'c2': from adapter notes: `{ JdbcDataType = 1 (CHAR), typeName = 'ENUM' }`, type = `CHAR(1) ASCII`.

Column descriptions from query `select * from T`, as provided by JDBC driver: `c1 CHAR(1) UTF8, c2 CHAR(1) UTF8`.

I will also enhance the integration tests of VSMYSQL to reproduce this output.

**Comment by djamadeus** [comment]
btw same some problem occurs with postgres adapter...
